install/kubernetes/cilium/values.yaml.tmpl

```diff
@@ -1409,18 +1392,12 @@ ipam:
   # -- Maximum rate at which the CiliumNode custom resource is updated.
   ciliumNodeUpdateRate: "15s"
   operator:
-    # -- Deprecated in favor of ipam.operator.clusterPoolIPv4PodCIDRList.
-    # IPv4 CIDR range to delegate to individual nodes for IPAM.
-    clusterPoolIPv4PodCIDR: "10.0.0.0/8"
     # -- IPv4 CIDR list range to delegate to individual nodes for IPAM.
-    clusterPoolIPv4PodCIDRList: []
+    clusterPoolIPv4PodCIDRList: ["10.0.0.0/8"]
     # -- IPv4 CIDR mask size to delegate to individual nodes for IPAM.
     clusterPoolIPv4MaskSize: 24
-    # -- Deprecated in favor of ipam.operator.clusterPoolIPv6PodCIDRList.
-    # IPv6 CIDR range to delegate to individual nodes for IPAM.
-    clusterPoolIPv6PodCIDR: "fd00::/104"
-     # -- IPv6 CIDR list range to delegate to individual nodes for IPAM.
-    clusterPoolIPv6PodCIDRList: []
+    # -- IPv6 CIDR list range to delegate to individual nodes for IPAM.
+    clusterPoolIPv6PodCIDRList: ["fd00::/104"]
```

Comment:
I'm paranoid that this will cause upgrade breakage. What kind of validation do we have for this new default to ensure that when users upgrade from earlier versions to this version, then new default doesn't either override their settings or change datapath behaviour?

Reply:
(I recognize that it's just moving a parameter from the non-list version to the list version, but this is the sort of innocent-looking change that can end up having an impact on an existing cluster during upgrade)

Reply:
I'm not aware of any specific validation that applies here :grimacing:. I've just tried to reason about the different cases:

- Users were setting `clusterPoolIPv*CIDR` but not the list versions: they'll have to switch since deprecated option is removed, they should see an error from the agent otherwise

- Users were setting both: same as above, they'll have to remove the one which is gone

- Users were setting `clusterPoolIPv*CIDRlist` already: their values will overwrite the new default and behaviour should be maintained

- Users were setting none: the default CIDRs are the same, but now taken from the list version

But I agree with you and don't feel super comfortable changing the default. I don't see an easier option, though: if I don't set this default value for `clusterPoolIPv*CIDRList`, we lose the default behaviour (and CI break).

Do you have a specific check in mind that we could add?

---

Regarding the current CI failures, they look related to the Hubble changes, but I haven't found the time to look into the details yet.

Reply:
> Users were setting clusterPoolIPv*CIDR but not the list versions: they'll have to switch since deprecated option is removed, they should see an error from the agent otherwise

Won't it just ignore the user setting and override with the list default?

Reply:
Maybe we can just say, "this is what upgrade release notes are for" and point to them if anything breaks.

Reply:
> Won't it just ignore the user setting and override with the list default?

Ah right, we're not removing any agent option here, so probably it will just ignore the unknown value and override with `clusterPoolIPv*CIDRList` (I'm finding [comments from Chris](https://github.com/helm/helm/issues/6422#issuecomment-1007179892) - looks like there's a PR open to make Helm lint fail when some values are not consumed, but it's not there yet). Hmm but I'm not really sure what's the best way to prevent that. :thinking: Should we force an error somehow if the legacy option is set? Do we want to open that box? (Or maybe we're already doing this for other, previously removed option? I don't know. I'm not super familiar with how we're usually handling this aspect of Helm charts.)

> Maybe we can just say, "this is what upgrade release notes are for" and point to them if anything breaks.

That's also true, and the documentation has marked the option as deprecated since v1.11, which left users with three minor version updates to adjust.

Not sure what's the best way. Let me know if you feel we need to trigger an error if the old value remains set.

Reply:
> Should we force an error somehow if the legacy option is set? 

I like this idea, bring up the failure before even upgrading / installing rather than failing quietly after install.

Reply:
I'm not sure how to do that, but I can give it a try next week. [It looks like](https://austindewey.com/2018/12/28/helm-tricks-input-validation-with-required-and-fail/) there's a `fail` directive available.

I suppose you prefer to have this in the same PR?

Reply:
Something like this maybe (not tested yet)

```diff
diff --git a/install/kubernetes/cilium/templates/cilium-configmap.yaml b/install/kubernetes/cilium/templates/cilium-configmap.yaml
index da5854d2a81a..cc6d337d29ed 100644
--- a/install/kubernetes/cilium/templates/cilium-configmap.yaml
+++ b/install/kubernetes/cilium/templates/cilium-configmap.yaml
@@ -814,10 +814,16 @@ data:
 
 {{- if or (eq $ipam "cluster-pool") (eq $ipam "cluster-pool-v2beta") }}
 {{- if .Values.ipv4.enabled }}
+  {{- if hasKey .Values.ipam.operator "clusterPoolIPv4PodCIDR" }}
+    {{- fail "Value ipam.operator.clusterPoolIPv4PodCIDR removed, use .clusterPoolIPv4PodCIDRList instead" }}
+  {{- end }}
   cluster-pool-ipv4-cidr: {{ .Values.ipam.operator.clusterPoolIPv4PodCIDRList | join " " | quote }}
   cluster-pool-ipv4-mask-size: {{ .Values.ipam.operator.clusterPoolIPv4MaskSize | quote }}
 {{- end }}
 {{- if .Values.ipv6.enabled }}
+  {{- if hasKey .Values.ipam.operator "clusterPoolIPv6PodCIDR" }}
+    {{- fail "Value ipam.operator.clusterPoolIPv6PodCIDR removed, use .clusterPoolIPv6PodCIDRList instead" }}
+  {{- end }}
   cluster-pool-ipv6-cidr: {{ .Values.ipam.operator.clusterPoolIPv6PodCIDRList | join " " | quote }}
   cluster-pool-ipv6-mask-size: {{ .Values.ipam.operator.clusterPoolIPv6MaskSize | quote }}
 {{- end }}
```

Reply:
It might also be possible to (mis?)-use the `upgradeCompatibility` Helm value and have different behavior based on users who are indicating (with that flag) that they are upgrading from v1.13 and older vs users who install from scratch.

Reply:
@gandro Sorry I don't follow you, what different behaviours would you implement with that variable?

Reply:
Apologies, after thinking about this some more, it doesn't make sense. The idea was to accept the deprecated flag if `upgradeCompatibility` - but that means we can basically never remove it. 

So the current behavior of failing hard when it has been mistakenly set is likely more sustainable.

Reply:
FWIW, I do think the approach with `fail` above is probably a good idea to keep around for at least one or two point releases. The cluster pool CIDR is a rather fundamental setting that cannot be changed (easily) after installation, so it might failing hard in Helm if the user is attempting to set it using the deprecated option.

Reply:
OK, thank you for the feedback! I'll try to validate the above and finalise the PR sometime this week.